### DISCUSSION TITLE
Parallelism in dynamic view diagrams

### DISCUSSION THREAD

**Discussion by susliko** [likec4/likec4#816] [Ideas]
@davydkov Thank you for the amazing dynamic diagrams feature!

Could you please share your plans on further evolution for it?

In particular, i'd love it to have a way to define interactions which are happening in parallel.
For example, by introducing some `parallel` block (possibly nested):

```
dynamic view X {
   parallel {
     A -> B // 1
     A -> C // 1
     parallel {
       C -> D // 2
       C -> E // 2
     }
   }
   A -> D // 3
}
```

And then to assign the same sequence number to all relations at the same nesting level.

If that's a viable syntax and semantics, I'll try contributing it 🙂

**Comment by davydkov** [comment]
Hey @susliko, @anatoly-kryzhanovsky 

>Could you please share your plans on further evolution for it?

At the moment, I am considering the idea of zooming in/out for activities (relationships/interactions).
It is coming from the #218 by @ThePlenkov

In a nutshell:
We drill down to the element's hierarchies.
Why don't we do the same for relationships?

For example, some high-level dynamic view:

```zig
A -> B ‘does something’
```

This activity can be detailed further, for example:

```zig
A.NestedA -> B.NestedB1 ‘calls api1’
A.NestedA <- B.NestedB1
A.NestedA -> B.NestedB2 ‘calls api2’
A.NestedA -> C ‘another api not relevant for high-level’
```

Click on `A -> B` from the high level may lead to the detailed view, which explains what is actually happening inside.

However, several questions are coming, such as:

- Is this only for dynamic views? Should it be something like "relationship views" ?
- What to do with returns `A <- B`?
- Do we need to validate that detailed view indeed contains `A.* -> B.*`?

I'm looking forward to your thoughts/ideas.

**Reply by susliko** [reply]
Zooming in/out for activities would be a killer feature 🚀 

I can think of the following use case for it.
Imagine we have a bunch of business processes. In terms of Big Bank Example these would be:
* view account balances
* make payments

Let's say we've visualized each of them using dynamic view diagrams.

Now we're looking at a Landscape diagram and want to answer the question "Which business processes make up this relationship between Internet Banking System and Mainframe Banking System?". 

Clicking on this relationship might open a selector with 2 diagrams of our business processes.

> Should it be something like "relationship views" ?

Sorry, I couldn't grasp the concept of relationship views in this context

> What to do with returns A <- B?

Do we need returns at all? Isn't enough that we can write it as  `B -> A`?

**Comment by davydkov** [comment]
> For example, by introducing some parallel block (possibly nested):

I have a question regarding this example.

If block/group

```zig
parallel {
  A -> B  // 1
  A -> C  // 1
  parallel {
    C -> D // 2
    C -> E // 2
  }
}
```

implies parallel interactions within the group,
then `A -> B` and `A -> C` should be parallel with nested `parallel`
(all inside same group).

Which means:

```zig
parallel {
  A -> B  // 1
  A -> C  // 1
  parallel {
    C -> D // 1
    C -> E // 1
  }
}
```

Or better:

```zig
parallel {
  A -> B  // 1.1
  A -> C  // 1.2
  parallel {
    C -> D // 1.3.1
    C -> E // 1.3.2
  }
}

```

where the common prefix indicates a common time frame, i.e., all `1.x` are parallel, and `1.3.x` too.

> by the way, nested `parallel` seems associative

If I understood you correctly, you'd instead like:

```zig
parallel {
  A -> B  // 1.1
  A -> C  // 1.2
}
parallel {
  C -> D // 2.1
  C -> E // 2.2
}
```

where `2.x` happens after `1.2` (without waiting for `1.1`).

But how do we define this?
First, what comes to mind:

```zig
// nested
parallel {
  A -> B
  A -> C {
    parallel {
      C -> D
      C -> E
    }
  }
}
```

But I think, eventually, this may lead to spaghetti.


Or something like time-labels (barriers):

```zig
parallel {
  A -> B '….' as @act1
  A -> C '…' as @act2
}

after @act2 {
  parallel {
    C -> D
    C -> E
  }
} as @act3

// maybe keywords: on, when ... ?
on (@act1, @act3) {
}
```

This seems more attractive (as it allows describing, for example, failure cases), but I think it needs further and careful research.

> [!NOTE]
> For example, might be no need in `parallel` keyword
>

```zig
 A -> B '….' as @act1
 
 // act2 after act1
 after @act1 {
   B -> C
 } as @act2
 
 // act3 also after act1 (i.e. parallel with act2)
 after @act1 {
   B -> D
 } as @act3
 
 // when act2 and act3 completed
 after (@act2, @act3) {
 }
```

Or some combination of both, or something else?

Thoughts?

**Comment by davydkov** [comment]
As a good first option, and probably most sufficient for now - do not allow nested parallel.

@susliko What do you think?

Contributions are more than welcome, but I think it is better if we do a code introduction call.
Or I update (better say "create") contribution guides.

**Comment by susliko** [comment]
@davydkov Thank you for such a thorough response!

My initial example was somewhat haphazard. Indeed, nested parallel blocks that I proposed just wouldn't make sense as 
```plantuml
   parallel {
     A -> B 
     A -> C 
     parallel {
       C -> D 
       C -> E 
     }
   }
```
is equivalent to 
```plantuml
   parallel {
     A -> B 
     A -> C 
     C -> D 
     C -> E 
   }
```
as you have mentioned.



> If I understood you correctly, you'd instead like:
...
where 2.x happens after 1.2 (without waiting for 1.1).

This scenario is quite complex, so that it's not even obvious how to mark and visualize relationships that start running in a background. 

What I actually meant was a structured-concurrency-like flow, when a parallel block must complete before a next step is allowed to execute.

This case is simple and can be covered without nested parallel blocks like you've mentioned:
```plantuml
parallel {
  A -> B  // 1
  A -> C  // 1
}
parallel {
  C -> D // 2
  C -> E // 2
}
```

> I think it is better if we do a code introduction call.
Or I update (better say "create") contribution guides.

Any of that would be great!

**Reply by davydkov** [reply]
@susliko What do you think?

Preview docs:
https://3fc5dadf.likec4-docs-astro.pages.dev/dsl/dynamic-views/#parallel

Preview playground:
https://af642fad.likec4-playground.pages.dev/w/dynamic/

**Reply by susliko** [reply]
@davydkov Looks great!
Just one small thing. I don't know if the current parser have such error-reporting capabilities, but it would be great to give a diagnostic that nested parallel blocks just don't make sense and are disallowed.
![image](https://github.com/user-attachments/assets/9c84dbad-d82d-401c-ae35-07decaa18f4c)

**Comment by johnny** [comment]
Hi all,

I recently started using dynamic views and here are my thoughts.

First, I was surprised, that relationships in dynamic views cannot contain all properties. I was trying to use the `link` property to achieve something similar to the "zooming in" described above.

Second, zooming in seems nice for simpler interactions, e.g. one more hop. But for more complex interactions I fear it becomes too rigid and might actually require duplicate work. A link to another graph could be an option. How would zoom work for export?

Third, I would find it useful to have the zoom feature available in all graphs. E.g. A relationship of type "broker" could be expanded to show details of the broker hop (topic name, ....)

**Reply by davydkov** [reply]
Hey @johnny 

In v1.10 it is possible to zoom in the relationships

<img width="582" alt="image" src="https://github.com/user-attachments/assets/807fee34-d0ef-482e-9f62-8ab42f70318e">

By setting `navigateTo` ([docs](https://likec4.dev/dsl/relationships/#navigate-to))
```zig
model {
  webApp -> backend.api {
    title 'requests data for the dashboard'
    navigateTo dashboardRequestFlow
  }
}
```
 
What do you think?

**Reply by johnny** [reply]
Hey @davydkov , the navigateTo is very nice. I just tried it out and the diagrams become even more magical.

Thank you for being so responsive. I haven't even merged my documentation changes, yet 😀 .

Could you also make the linking in the elements take links to other diagrams?

**Reply by anujkumar-df** [reply]
Hi @davydkov is navigateTo only applicable at model?

**Reply by davydkov** [reply]
Hey @anujkumar-df 
I'm not sure I understood the question. Can you give an example where you want to apply it?

**Reply by anujkumar-df** [reply]
Please ignore, it was working, I was just using an incorrect syntax.

**Comment by simonwfarrow** [comment]
Loving the parallel blocks, problem is, the more you add the more we want! I was wondering if an 'optional' statement was possible? along a similar vain as parallel, I would like to shop steps that were optional

**Reply by elf-pavlik** [reply]
optional would be great, I'm very used to it from http://mermaid.js.org/syntax/sequenceDiagram.html#alt

**Comment by Yash-xoxo** [comment]
### The Problem

I completely understand the frustration here. We have a large codebase that's migrating from JavaScript to TypeScript, and the "all or nothing" nature of TypeScript's type system creates real friction:

- Legacy modules that would take weeks to properly type
- Third-party libraries with incomplete or incorrect type definitions  
- Prototyping new features where types would slow down iteration
- Integration points with dynamic data sources

Right now, the options are all suboptimal:
- `any` - disables all type checking, defeats the purpose
- `unknown` - too restrictive, requires constant type guards
- `// @ts-ignore` - feels like giving up, hides real issues
- Proper typing - sometimes just not feasible on the timeline

### What I'd Love to See

Something like:
```typescript
// Opt-in per variable/function
let data: dynamic = fetchFromAPI();
data.whatever.deeply.nested; // No error, checked at runtime

// Or per-file gradual mode
// @ts-gradual
function processLegacyData(input) {
  // TypeScript provides hints but doesn't block
  return input.transform().map(x => x.value);
}
```

**Key requirements:**
- Runtime safety (throw on undefined access, not silent failures)
- IDE support (autocomplete from inferred types when possible)
- Explicit opt-in (don't break existing strict code)
- Clear boundaries (know where gradual typing starts/stops)

### Concerns I Have

That said, I'm sympathetic to the TypeScript team's position:

**1. Complexity:** Adding a third type system (strict, gradual, `any`) could make the language harder to learn and maintain.

**2. Runtime overhead:** True gradual typing needs runtime checks. Who ships that? Does it get stripped in production? What about performance?

**3. Escape hatch abuse:** Would people use this instead of properly typing code, creating technical debt?

**4. Two worlds problem:** Projects might split into "gradual" and "strict" codebases that don't interoperate well.

### Practical Alternatives Today

While we wait (or if this never happens), here are some patterns that help:

**1. Type assertion with runtime validation:**
```typescript
import { z } from 'zod';

const UserSchema = z.object({
  name: z.string(),
  age: z.number(),
});

const data = UserSchema.parse(apiResponse); // Runtime check + type inference
```

**2. Branded types for external data:**
```typescript
type Unvalidated<T> = T & { __unvalidated: true };
type Validated<T> = T;

function validate<T>(data: Unvalidated<T>): Validated<T> {
  // runtime checks
  return data as Validated<T>;
}
```

**3. Progressive enhancement:**
```typescript
// Start with loose types
type LegacyData = Record<string, unknown>;

// Gradually tighten as you understand the shape
type LegacyData = {
  id: string;
  [key: string]: unknown; // Still flexible but safer
};
```

### My Take

I think TypeScript's "strict or bust" philosophy has been key to its success - it avoids the messy middle ground that makes some gradually-typed languages hard to reason about. But I also feel the pain of the migration path.

Maybe the solution isn't in TypeScript itself, but in better tooling:
- Automated type generation from runtime data
- Better incremental migration tools
- Smarter inference for common patterns
- Standard library for runtime validation with type inference

Would love to hear from others dealing with this in large codebases. How are you handling the migration from untyped/loosely-typed code?

---

**Side note:** For anyone interested in the academic side, Typed Racket and Pyright's approach to gradual typing might be worth studying for ideas.
